### PR TITLE
Minor Shader Graph fixes #2

### PR DESCRIPTION
**Pull request by Kink3d**
### Purpose of this PR
Minor bugfixes again.

- Removed some pointless double delimiters
- Added a callback to Exposed toggle of blackboard properties to enable/disable other parameters
- Fixed a bunch of issues with blackboard vector1 fields and mode switches
- Rewrite Slider node UI callbacks to fix a few bugs
- Add a bunch of missing undo stack actions
- Sanitise new property names when converting from inline
- Add ShowGeneratedCode/CopyShader back to Master nodes
- Return Master nodes to 200 width
- Hardcode ColorField width

Fixes the following Fogbugz cases:
https://fogbugz.unity3d.com/f/cases/1091510/
https://fogbugz.unity3d.com/f/cases/1091503/
https://fogbugz.unity3d.com/f/cases/1091495/
https://fogbugz.unity3d.com/f/cases/1091496/
https://fogbugz.unity3d.com/f/cases/1091497/
https://fogbugz.unity3d.com/f/cases/1091494/
https://fogbugz.unity3d.com/f/cases/1089978/
https://fogbugz.unity3d.com/f/cases/1089979/
https://fogbugz.unity3d.com/f/cases/1086696/
https://fogbugz.unity3d.com/f/cases/1086695/
https://fogbugz.unity3d.com/f/cases/1086694/

---
### Release Notes
Changelog...

---
### Testing status
**Katana Tests**: Green.

**Manual Tests**: @spaztronaut 

**Automated Tests**: None

Any test projects to go with this to help reviewers? N/A

---
### Overall Product Risks
**Technical Risk**: Medium (undo stack)

**Halo Effect**: Medium (undo stack)

---
### Comments to reviewers
@spaztronaut Please pay very close attention to the changes to the undo stack! This cannot be covered by automation and is pretty susceptible to regressions!

### CODE REVIEW

**Comment by pbbastian** [comment]
Can you write up a bit more about what this PR actually does? 11 fogbugz links are not super helpful, as I then have to identify what the bug actually was. This PR does touch quite a lot of code.

Also, I think forgot to commit the change log :)

**Comment by spaztronaut** [comment]
zack said he'd check this one out

File: com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardFieldPropertyView.cs [hunk 1]
```diff
@@ -250,18 +145,25 @@ public BlackboardFieldPropertyView(AbstractMaterialGraph graph, IShaderProperty
...
                 var defaultModeField = new EnumField((Enum)textureProperty.defaultType);
                 defaultModeField.OnValueChanged(evt =>
                     {
+                        m_Graph.owner.RegisterCompleteObjectUndo("Change Texture Mode");
                         if (textureProperty.defaultType == (TextureShaderProperty.DefaultType)evt.newValue)
                             return;
                         textureProperty.defaultType = (TextureShaderProperty.DefaultType)evt.newValue;
                         DirtyNodes(ModificationScope.Graph);
                     });
+                void toggleDefaultModeFieldEnabled()
```
Comment (Kink3d) [inline com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardFieldPropertyView.cs]: @pbbastian Thought I'd draw your attention here as idk if we are OK to use inline functions in the codebase now?

Comment (pbbastian) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardFieldPropertyView.cs]: That's fine, but please use the same naming conventions as normal methods :)

Comment (pbbastian) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardFieldPropertyView.cs]: In fact, they perform just as well as regular methods, and better than lambdas. (Unless you convert them to a lambda of course.)

**Comment by Kink3d** [comment]
Tests green against trunk

File: com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardFieldPropertyView.cs (hunk 1 above)
Comment (Kink3d) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardFieldPropertyView.cs]: Fixed

**Comment by Kink3d** [comment]
@pbbastian can I get another review on this please?

File: com.unity.shadergraph/CHANGELOG.md [hunk 2]
```diff
@@ -33,6 +33,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Non-exposed properties in the blackboard no longer have a green dot next to them.
 - Default reference name for shader properties are now serialized. You cannot change them after initial creation.
 - When you save Shader Graph and Sub Graph files, they're now automatically checked out on version control.
+- Converting an inline node to a `Property` node no longer allows duplicate property names.
```
Comment (Katstodian) [inline com.unity.shadergraph/CHANGELOG.md]: Please change to:
When you convert an inline node to a __Property__ node, this no longer allows duplicate property names.

File: com.unity.shadergraph/CHANGELOG.md [hunk 3]
```diff
@@ -33,6 +33,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- Moving a node now triggers a save needed action.
```
Comment (Katstodian) [inline com.unity.shadergraph/CHANGELOG.md]: Please change to:
When you move a node, you'll now be asked to save it.

(unless the "save needed action" means "SG now saves"?)

File: com.unity.shadergraph/CHANGELOG.md [hunk 4]
```diff
@@ -33,6 +33,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- Editing Property parameters on the Blackboard is now recorded for Undo actions.
```
Comment (Katstodian) [inline com.unity.shadergraph/CHANGELOG.md]: DOes this mean:
"You can now Undo edits to Property parameters on the Blackboard."?

File: com.unity.shadergraph/CHANGELOG.md [hunk 5]
```diff
@@ -33,6 +33,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- Converting a Property node to an inline node and vice versa is now recorded for Undo actions.
```
Comment (Katstodian) [inline com.unity.shadergraph/CHANGELOG.md]: Does this mean:
You can now Undo conversions between Property nodes and inline nodes. 
?

File: com.unity.shadergraph/CHANGELOG.md [hunk 6]
```diff
@@ -33,6 +33,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- Moving a node is now recorded for Undo actions.
```
Comment (Katstodian) [inline com.unity.shadergraph/CHANGELOG.md]: Does this mean:
You can now Undo moving a node.
?

File: com.unity.shadergraph/CHANGELOG.md [hunk 7]
```diff
@@ -33,6 +33,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- `Texture2D` Property type `Mode` dropdown is no longer selectable when the Property is not exposed. This previously had no effect.
```
Comment (Katstodian) [inline com.unity.shadergraph/CHANGELOG.md]: Please change to:
You can no longer select the __Texture2D__ Property type __Mode__, if the Property is not exposed.

File: com.unity.shadergraph/CHANGELOG.md [hunk 8]
```diff
@@ -33,6 +33,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- `Vector1` Property type now handles default values more intuitively when switching `Mode` dropdown.
```
Comment (Katstodian) [inline com.unity.shadergraph/CHANGELOG.md]: What does "handles more intuitively" mean here?

File: com.unity.shadergraph/CHANGELOG.md [hunk 9]
```diff
@@ -33,6 +33,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- `Color` node control is now a consistent width.
```
Comment (Katstodian) [inline com.unity.shadergraph/CHANGELOG.md]: Please replace with:
The __Color__ node control is now a consistent width.

(don't forget your articles :p)

File: com.unity.shadergraph/CHANGELOG.md [hunk 10]
```diff
@@ -33,6 +33,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- Function declarations no longer contain doubel delimiters.
```
Comment (Katstodian) [inline com.unity.shadergraph/CHANGELOG.md]: please replace "doubel" with "double"

(so close!)

File: com.unity.shadergraph/CHANGELOG.md [hunk 11]
```diff
@@ -33,6 +33,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- `Slider` node control now functions correctly.
```
Comment (Katstodian) [inline com.unity.shadergraph/CHANGELOG.md]: Please replace with:

The __Slider__ node control now works correctly.

File: com.unity.shadergraph/CHANGELOG.md (hunk 8 above)
Comment (Kink3d) [reply com.unity.shadergraph/CHANGELOG.md]: Ok this is the only one left as I didnt really know what to put here. Its actually a total rewrite of the UI for it. It had a number of problems that are tied together but pretty difficult to describe in one sentence.

Essentially it retains and resets the defaults when you switch modes in ways that are more akin to regular fields in the editor.

**Comment by Katstodian** [comment]
Why is there an .md file from the LWRP documentation in this PR? The Builtin/LWRP Feature comparison page?

**Comment by Kink3d** [comment]
The extra LWRP MD file was the result of a bad merge. It has now been removed.

**Pull request merged**